### ISSUE TITLE
Adding torch.Generator to WeightedRandomSampler to ensure pseudorandom results in development phase

### ISSUE THREAD

**Issue by TCACastelijns** [pytorch/pytorch#40274]
## 🚀 Feature
Adding torch.Generator to `WeightedRandomSampler` to ensure pseudorandom results in development phase

## Motivation
Currently, I am working with two collegues on a project where we are training, validating and testing neural networks on production data in a classic unbalanced classification problem. Since we have unbalance in our positive class, we are applying the `WeightedRandomSampler` to ensure balanced classes in training phase. In development we are also setting a random seed to ensure that we can reproduce our results and compare with the members of the team.

The problem is the current implementation of the `WeightedRandomSampler`. It does not contain a `random_state` argument or something similar. Therefore, if we use `WeightedRandomSampler` as preprocess step, we are not able to reproduce the same results. Especially in developing phase, this is a desired feature.  Therefore, we are suggesting the following feature:
Adding `generator` input argument to `WeightedRandomSampler` class and use it in the 
`torch.multinomial` implementation.

## Pitch
```
class WeightedRandomSampler(Sampler):

     """Samples elements from ``[0,..,len(weights)-1]`` with given probabilities (weights).

    Args:
        weights (sequence)   : a sequence of weights, not necessary summing up to one
        num_samples (int): number of samples to draw
        generator (torch.Generator): a pseudorandom number generator for sampling
        replacement (bool): if ``True``, samples are drawn with replacement.
            If not, they are drawn without replacement, which means that when a
            sample index is drawn for a row, it cannot be drawn again for that row.

    Example:
        >>> list(WeightedRandomSampler([0.1, 0.9, 0.4, 0.7, 3.0, 0.6], 5, replacement=True))
        [4, 4, 1, 4, 5]
        >>> list(WeightedRandomSampler([0.9, 0.4, 0.05, 0.2, 0.3, 0.1], 5, replacement=False))
        [0, 1, 4, 3, 2]
        >>> list(WeightedRandomSampler([0.9, 0.4, 0.05, 0.2, 0.3, 0.1], 5, torch.Generator(), replacement=False))
        [0, 4, 1, 3, 5]
    """

    def __init__(self, weights, num_samples, generator=None, replacement=True):
        if not isinstance(num_samples, _int_classes) or isinstance(num_samples, bool) or \
                num_samples <= 0:
            raise ValueError("num_samples should be a positive integer "
                             "value, but got num_samples={}".format(num_samples))
        if not isinstance(replacement, bool):
            raise ValueError("replacement should be a boolean value, but got "
                             "replacement={}".format(replacement))
        self.weights = torch.as_tensor(weights, dtype=torch.double)
        self.num_samples = num_samples
        self.generator = generator
        self.replacement = replacement


    def __iter__(self):
        return iter(
            torch.multinomial(self.weights, self.num_samples, self.generator, self.replacement).tolist())

    def __len__(self):
        return self.num_samples
```

cc @SsnL @pbelevich

**Comment by ezyang** [comment]
Accepting a generator sounds good, send in a PR.

**Comment by SsnL** [comment]
already implemented on master!
